# 2. Introduction to FastAPI
In this notebook, we will start building our own APIs using **FastAPI**.

## Introduction
**FastAPI** is a modern, fast (high-performance), web framework for building APIs with Python 3.6+ based on standard Python type hints.

### Why FastAPI?
- **Fast**: Very high performance, on par with NodeJS and Go (thanks to Starlette and Pydantic).
- **Fast to code**: Increase the speed to develop features by about 200% to 300%.
- **Fewer bugs**: Reduce about 40% of human (developer) induced errors.
- **Intuitive**: Great editor support. Completion everywhere. Less time debugging.
- **Easy**: Designed to be easy to use and learn. Less time reading docs.
- **Short**: Minimize code duplication. Multiple features from each parameter declaration.
- **Robust**: Get production-ready code. With automatic interactive documentation.
- **Standards-based**: Based on (and fully compatible with) the open standards for APIs: OpenAPI (previously known as Swagger) and JSON Schema.

### Useful Resources
- [FastAPI Documentation](https://fastapi.tiangolo.com/)
- [Pydantic Documentation](https://docs.pydantic.dev/latest/)
- [Starlette Documentation](https://www.starlette.io/)

In [2]:
# Imports
from fastapi import FastAPI
from fastapi.testclient import TestClient

## 2.1 Learning Goals
- Create a basic FastAPI application.
- Understand the `@app.get` decorator.
- Use `TestClient` to test the API within the notebook.
- Understand Path Parameters and Query Parameters in FastAPI.
- Learn about **Swagger UI** (Automatic Documentation).

### Notebook roadmap
- Build a tiny API (`/`) to see the FastAPI + `TestClient` loop.
- Explore auto-docs (Swagger UI) conceptually even though we run in-notebook.
- Add path params (`/items/{item_id}`) and see validation errors FastAPI gives you for free.
- Add query params (`/users/`) and optional defaults.
- Finish with a small exercise to practice combining both kinds of parameters.
Tip: because the `app` object is reused, run cells in order so earlier routes are defined before later requests.


## 2.2 Your First API
We will create a simple API with a single endpoint that returns a "Hello World" message.

In [3]:
# 1. Create the FastAPI app instance
app = FastAPI()

# 2. Define a path operation decorator
@app.get("/")
def read_root():
    return {"message": "Hello World"}

# 3. Create a TestClient to interact with the app
client = TestClient(app)

# 4. Make a request to the API
response = client.get("/")
print(response.status_code)
print(response.json())

200
{'message': 'Hello World'}


## 2.3 Automatic Documentation (Swagger UI)
One of the best features of FastAPI is that it automatically generates interactive API documentation for you.

If you were running this code in a terminal (using `uvicorn main:app --reload`), you could visit `http://127.0.0.1:8000/docs` in your browser to see the **Swagger UI**.

Swagger UI allows you to:
- See all available endpoints.
- See what parameters they expect.
- Execute requests directly from the browser.

Since we are in a notebook, we will use `TestClient` to simulate requests, but keep in mind that in a real project, Swagger UI is your best friend for testing and documentation.

## 2.4 Path Parameters
Path parameters are parts of the URL path that are variable.
Example: `/items/5` where `5` is the `item_id`.

In [4]:
@app.get("/items/{item_id}")
def read_item(item_id: int):
    return {"item_id": item_id}

# Test it
response = client.get("/items/42")
print(response.json())

# Test with invalid type (string instead of int)
response = client.get("/items/foo")
print(response.status_code)
print(response.json())

{'item_id': 42}
422
{'detail': [{'type': 'int_parsing', 'loc': ['path', 'item_id'], 'msg': 'Input should be a valid integer, unable to parse string as an integer', 'input': 'foo'}]}


## 2.5 Query Parameters
Function parameters that are not part of the path are automatically interpreted as "query" parameters.

In [5]:
@app.get("/users/")
def read_user(user_id: int, q: str = None):
    return {"user_id": user_id, "q": q}

# Test with query param
response = client.get("/users/?user_id=1&q=somequery")
print(response.json())

# Test without optional query param
response = client.get("/users/?user_id=1")
print(response.json())

{'user_id': 1, 'q': 'somequery'}
{'user_id': 1, 'q': None}


## 2.6 Exercise: Personalized Greeting
**Task**:
1. Create a new endpoint `/greet/{name}`.
2. It should take `name` as a path parameter.
3. It should take an optional query parameter `is_formal` (boolean, default `False`).
4. If `is_formal` is `True`, return `{"message": "Good day, {name}"}`.
5. Otherwise, return `{"message": "Hi, {name}"}`.

In [5]:
# TODO: Write your code here

### Solution

In [6]:
# Solution
@app.get("/greet/{name}")
def greet(name: str, is_formal: bool = False):
    if is_formal:
        return {"message": f"Good day, {name}"}
    else:
        return {"message": f"Hi, {name}"}

# Test cases
print(client.get("/greet/Bob").json())
print(client.get("/greet/Bob?is_formal=true").json())

{'message': 'Hi, Bob'}
{'message': 'Good day, Bob'}


## 2.1 Learning Goals
- Create a basic FastAPI application.
- Understand the `@app.get` decorator.
- Use `TestClient` to test the API within the notebook.
- Understand Path Parameters and Query Parameters in FastAPI.
- Learn about **Swagger UI** (Automatic Documentation).

## 2.2 Your First API
We will create a simple API with a single endpoint that returns a "Hello World" message.

In [7]:
# 1. Create the FastAPI app instance
app = FastAPI()

# 2. Define a path operation decorator
@app.get("/")
def read_root():
    return {"message": "Hello World"}

# 3. Create a TestClient to interact with the app
client = TestClient(app)

# 4. Make a request to the API
response = client.get("/")
print(response.status_code)
print(response.json())

200
{'message': 'Hello World'}


## 2.3 Automatic Documentation (Swagger UI)
One of the best features of FastAPI is that it automatically generates interactive API documentation for you.

If you were running this code in a terminal (using `uvicorn main:app --reload`), you could visit `http://127.0.0.1:8000/docs` in your browser to see the **Swagger UI**.

Swagger UI allows you to:
- See all available endpoints.
- See what parameters they expect.
- Execute requests directly from the browser.

Since we are in a notebook, we will use `TestClient` to simulate requests, but keep in mind that in a real project, Swagger UI is your best friend for testing and documentation.

## 2.4 Path Parameters
Path parameters are parts of the URL path that are variable.
Example: `/items/5` where `5` is the `item_id`.

In [8]:
@app.get("/items/{item_id}")
def read_item(item_id: int):
    return {"item_id": item_id}

# Test it
response = client.get("/items/42")
print(response.json())

# Test with invalid type (string instead of int)
response = client.get("/items/foo")
print(response.status_code)
print(response.json())

{'item_id': 42}
422
{'detail': [{'type': 'int_parsing', 'loc': ['path', 'item_id'], 'msg': 'Input should be a valid integer, unable to parse string as an integer', 'input': 'foo'}]}


## 2.5 Query Parameters
Function parameters that are not part of the path are automatically interpreted as "query" parameters.

In [9]:
@app.get("/users/")
def read_user(user_id: int, q: str = None):
    return {"user_id": user_id, "q": q}

# Test with query param
response = client.get("/users/?user_id=1&q=somequery")
print(response.json())

# Test without optional query param
response = client.get("/users/?user_id=1")
print(response.json())

{'user_id': 1, 'q': 'somequery'}
{'user_id': 1, 'q': None}


## 2.6 Exercise: Personalized Greeting
**Task**:
1. Create a new endpoint `/greet/{name}`.
2. It should take `name` as a path parameter.
3. It should take an optional query parameter `is_formal` (boolean, default `False`).
4. If `is_formal` is `True`, return `{"message": "Good day, {name}"}`.
5. Otherwise, return `{"message": "Hi, {name}"}`.

In [12]:
# TODO: Write your code here
@app.get("/greet/{name}")
def greet(name: str, is_formal: bool = False):
    if is_formal:
        return {"message": f"Good day, {name}"}
    else:
        return {"message": f"Hi, {name}"}

# Test cases
print(client.get("/greet/Bob").json())
print(client.get("/greet/Bob?is_formal=true").json())

{'message': 'Hi, Bob'}
{'message': 'Good day, Bob'}


## 2.1 Learning Goals
- Practice making GET requests to retrieve lists of data.
- Practice filtering data received from an API.
- Practice handling errors (e.g., 404 Not Found).

## Exercise 1: Fetch a List of Users
**Task**: 
1. Make a `GET` request to `https://jsonplaceholder.typicode.com/users`.
2. Check if the status code is 200.
3. Print the number of users returned.
4. Print the name of the first user in the list.

In [11]:
# TODO: Write your code here
url = "https://jsonplaceholder.typicode.com/users"

# 1. Make the request

# 2. Check status code

# 3. Print number of users

# 4. Print first user's name

## Exercise 2: Filter Data
**Task**:
1. Iterate through the list of users you fetched in Exercise 1.
2. Find and print the `email` of the user with the `username` "Samantha".

In [12]:
# TODO: Write your code here


## Exercise 3: Handling Errors
**Task**:
1. Try to make a `GET` request to a non-existent URL: `https://jsonplaceholder.typicode.com/users/9999`.
2. Check the status code.
3. Write an `if` statement that prints "User not found" if the status code is 404, and "User found" otherwise.

In [13]:
# TODO: Write your code here
invalid_url = "https://jsonplaceholder.typicode.com/users/9999"


## Wrap-up and next steps
- Try adding response models with Pydantic (`response_model=...`) to shape outputs.
- Add status codes to decorators (e.g., `@app.get(..., status_code=200)`) to make intent explicit.
- Explore dependency injection (`Depends`) to share auth or db clients across routes.
- Spin up the same app with `uvicorn` to see the real Swagger UI at `/docs` and `/redoc`.
- Think about error handling: what should a 404 vs 422 vs 500 look like for your API?
